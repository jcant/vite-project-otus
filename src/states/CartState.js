import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useCartStore = defineStore("cart", () => {
  const cart = ref([]); //array of objects: {count: number, product: object}

  function addToCart(newProduct) {
    if (newProduct != null) {
      let isProductPresent = false;
      cart.value.forEach((cartItem) => {
        if (cartItem.product.id == newProduct.id) {
          cartItem.count++;
          isProductPresent = true;
        }
      });
      if (!isProductPresent) {
        cart.value.push({ count: 1, product: newProduct });
      }
    }
  }

  function incProduct(id) {
    cart.value.forEach((cartItem) => {
      if (cartItem.product.id == id) {
        cartItem.count++;
      }
    });
  }

  function decProduct(id) {
    cart.value.forEach((cartItem) => {
      if (cartItem.product.id == id && cartItem.count > 1) {
        cartItem.count--;
      }
    });
  }

  function removeProduct(id) {
    for (let i = 0; i < cart.value.length; i++) {
      if (cart.value[i].product.id == id) {
        cart.value.splice(i, 1);
        break;
      }
    }
  }

  function $reset() {
    cart.value = [];
  }

  const count = computed(() => {
    let count = 0;
    cart.value.forEach((cartItem) => (count += cartItem.count));
    return count;
  });

  const totalPrice = computed(() => {
    let count = 0;
    cart.value.forEach(
      (cartItem) => (count += cartItem.count * cartItem.product.price)
    );
    return count;
  });

  return {
    $reset,
    cart,
    count,
    totalPrice,
    addToCart,
    incProduct,
    decProduct,
    removeProduct,
  };
});
