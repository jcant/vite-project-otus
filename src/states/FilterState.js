import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useFilterStore = defineStore("filter", () => {
  const filter = ref("");

  const isFiltered = computed(() => filter.value != "");

  return { filter, isFiltered };
});
