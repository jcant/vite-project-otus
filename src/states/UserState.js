import { ref, computed } from "vue";
import { defineStore } from "pinia";

export const useUserStore = defineStore("user", () => {
  const user = ref("");

  function login(username, password) {
    if (username != null && password != null) {
      user.value = username;
      saveToLocalStorage();
    }
  }

  function saveToLocalStorage() {
    if (isAuthorized) {
      localStorage.setItem("auth_user", user.value);
    }
  }

  function initFromLocalStorage() {
    user.value = localStorage.getItem("auth_user");
  }

  function logout() {
    user.value = "";
    localStorage.removeItem("auth_user");
  }

  function $reset() {
    logout();
  }

  const isAuthorized = computed(() => user.value != "" && user.value != null);

  return {
    user,
    isAuthorized,
    login,
    logout,
    initFromLocalStorage,
    $reset,
  };
});
