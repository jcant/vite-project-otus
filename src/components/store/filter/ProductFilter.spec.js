import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import component from "./ProductFilter.vue";
import { setActivePinia, createPinia, storeToRefs } from "pinia";
import { useFilterStore } from "@/states/FilterState";

describe("Product filter component", () => {
  let wrapper;

  beforeEach(() => {
    setActivePinia(createPinia());
    wrapper = mount(component);
  });

  it("imports", () => {
    expect(component).toBeDefined();
  });

  it("mounts without errors", () => {
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });
  it("renders all", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("set filter input control", () => {
    const filterStore = useFilterStore();
    const { filter, isFiltered } = storeToRefs(filterStore);
    const inputControl = wrapper.find('[data-testid="filter-input"]');

    expect(inputControl.exists()).toBe(true);
    inputControl.setValue("filter value");

    expect(isFiltered.value).toBe(true);
    expect(filter.value).toBe("filter value");
  });
  it("clear filter by click at clear button", () => {
    const filterStore = useFilterStore();
    const { filter, isFiltered } = storeToRefs(filterStore);
    filter.value = "some filter";
    wrapper = mount(component);
    const clearControl = wrapper.find('[data-testid="clear-filter"]');

    expect(clearControl.exists()).toBe(true);
    clearControl.trigger("click");

    expect(isFiltered.value).toBe(false);
    expect(filter.value).toBe("");
  });
});
