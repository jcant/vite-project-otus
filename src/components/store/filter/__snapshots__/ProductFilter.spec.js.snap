// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product filter component > renders all 1`] = `
"<div>
  <div><label>Search by price or by title</label>
    <div class="flex place-items-center cursor-pointer"><input data-testid="filter-input" type="text" class="border border-amber-800 rounded-2xl pl-2 pr-2 p-1"><img src="/src/assets/search.png" class="w-6 h-6 ml-2"></div>
  </div>
  <!--v-if-->
</div>"
`;
